src/mapstorage/jsonmapstorage.cpp

```diff
@@ -0,0 +1,589 @@
+/************************************************************************
+**
+** Authors:   Ulf Hermann <ulfonk_mennhar@gmx.de> (Alve),
+**            Marek Krejza <krejza@gmail.com> (Caligor),
+**            Nils Schimmelmann <nschimme@gmail.com> (Jahara)
+**
+** This file is part of the MMapper project.
+** Maintained by Nils Schimmelmann <nschimme@gmail.com>
+**
+** This program is free software; you can redistribute it and/or
+** modify it under the terms of the GNU General Public License
+** as published by the Free Software Foundation; either version 2
+** of the License, or (at your option) any later version.
+**
+** This program is distributed in the hope that it will be useful,
+** but WITHOUT ANY WARRANTY; without even the implied warranty of
+** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+** GNU General Public License for more details.
+**
+** You should have received a copy of the GNU General Public License
+** along with this program; if not, write to the:
+** Free Software Foundation, Inc.
+** 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+**
+************************************************************************/
+
+#include "jsonmapstorage.h"
+#include "mapdata.h"
+#include "roomsaver.h"
+#include "patterns.h"
+#include "configuration.h"
+#include "oldconnection.h"
+#include "oldroom.h"
+#include "mmapper2room.h"
+#include "mmapper2exit.h"
+#include "progresscounter.h"
+#include "basemapsavefilter.h"
+#include "infomark.h"
+#include "qtiocompressor.h"
+#include "olddoor.h"
+#include "parserutils.h"
+
+#include <QFile>
+#include <QDir>
+#include <QTextStream>
+#include <QFile>
+#include <QJsonArray>
+#include <QJsonObject>
+#include <QJsonDocument>
+#include <QMap>
+#include <QMultiMap>
+#include <QCryptographicHash>
+#include <QRegularExpression>
+
+#include <cassert>
+#include <iostream>
+
+using namespace std;
+
+namespace
+{
+
+typedef QList<const Room *> RoomList2;
```

Comment:
Did you have a major reason to not re-use `RoomList`? I know it's a vector rather than a QList. Alternative we can also rename the current `RoomList` -> `RoomVector`.

Reply:
I think I wanted QList behavior for some reason. I'll review it and see if it could use that QVector type instead.

Reply:
It can't be RoomList because in one instance it stores `const Room*`, not `Room *`, and in the other it needs to be compatible with `RoomSaver`.

In `json4` I clarify `RoomVector` (as you suggested) vs. `ConstRoomList`.